File: FinanceApp.Shared/Models/UserTables/Spending.cs
﻿using FinanceApp.Shared.Enum;
using FinanceApp.Shared.Models.UserTables.Bases;
using System.ComponentModel.DataAnnotations;

namespace FinanceApp.Shared.Models.UserTables
{
    public class Spending : UserTable
    {
        [Required]
        public string Name { get; set; }
        [Required]
        public ERecurrence Recurrence { get; set; }
        public decimal Amount { get; set; }
        public DateTime InitialDate { get; set; }
        public DateTime? EndDate { get; set; }
        public bool IsEndless { get; set; }
        public int? TimesRecurrence { get; set; }
        public bool IsRequired { get; set; }
        public Category? Category { get; set; }
        public EPayment Payment { get; set; }
        public CreditCard? CreditCard { get; set; } // Cartão de Crédito
    }
}

File: FlutterApp/finance_app/lib/models/spending/spending.dart
import 'package:finance_app/models/category/category.dart';
import 'package:finance_app/models/credit_card/credit_card.dart';

class Spending {
  int id;
  String name;
  int recurrence;
  int payment;
  CreditCard? creditCard;
  String recurrenceDisplayName;
  String paymentName;
  Category? category;
  bool isRequired;
  double amount;
  DateTime initialDate;
  DateTime? endDate;
  bool isEndless;
  int timesRecurrence;

  Spending(
      {required this.id,
      required this.name,
      required this.recurrence,
      required this.recurrenceDisplayName,
      required this.creditCard,
      required this.paymentName,
      required this.payment,
      required this.isRequired,
      required this.category,
      required this.amount,
      required this.initialDate,
      required this.endDate,
      required this.isEndless,
      required this.timesRecurrence});

  Spending.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        name = json['name'].toString(),
        recurrence = json['recurrence'],
        recurrenceDisplayName = json['recurrenceDisplayName'].toString(),
        amount = json['amount'].toDouble(),
        initialDate = DateTime.tryParse(json['initialDate'].toString())!,
        endDate = DateTime.tryParse(json['endDate'].toString()),
        isEndless = json['isEndless'],
        timesRecurrence = json['timesRecurrence'],
        category = json['category'] == null
            ? null
            : Category.fromJson(json['category']),
        creditCard = json['creditCard'] == null
            ? null
            : CreditCard.fromJson(json['creditCard']),
        isRequired = json['isRequired'],
        payment = json['payment'],
        paymentName = json['paymentName'];
}
